enclave-runtime/src/top_pool_execution.rs

```diff
@@ -0,0 +1,180 @@
+
+	let shards = state_handler.list_shards()?;
+	let mut remaining_shards = shards.len() as u32;
+	let ends_at = duration_now() + MAX_TRUSTED_GETTERS_EXEC_DURATION;
+
+	let top_pool_executor = TopPoolOperationExecutor::<Block, SignedSidechainBlock, _, _>::new(
+		rpc_author,
+		stf_executor,
+	);
+
+	// Execute trusted getters for each shard. Each shard gets equal amount of time to execute
+	// getters.
+	for shard in shards.into_iter() {
+		let shard_exec_time = match remaining_time(ends_at)
+			.map(|r| r.checked_div(remaining_shards))
+			.flatten()
+		{
+			Some(t) => t,
+			None => {
+				info!("[Enclave] Could not execute trusted operations for all shards. Remaining number of shards: {}.", remaining_shards);
+				break
+			},
+		};
+
+		match top_pool_executor.execute_trusted_getters_on_shard(&shard, shard_exec_time) {
+			Ok(()) => {},
+			Err(e) => error!("Error in trusted getter execution for shard {:?}: {:?}", shard, e),
+		}
+
+		remaining_shards -= 1;
+	}
+
+	Ok(())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn execute_trusted_calls() -> sgx_status_t {
+	if let Err(e) = execute_top_pool_trusted_calls_internal::<Block>() {
+		return e.into()
+	}
+
+	sgx_status_t::SGX_SUCCESS
+}
+
+/// Internal [`execute_trusted_calls`] function to be able to use the `?` operator.
+///
+/// Executes `Aura::on_slot() for `slot` if it is this enclave's `Slot`.
+///
+/// This function makes an ocall that does the following:
+///
+/// *   sends sidechain `confirm_block` xt's with the produced sidechain blocks
+/// *   gossip produced sidechain blocks to peer validateers.
+fn execute_top_pool_trusted_calls_internal<PB>() -> Result<()>
+where
+	PB: BlockT<Hash = H256>,
+	NumberFor<PB>: BlockNumberOps,
+{
+	// we acquire lock explicitly (variable binding), since '_' will drop the lock after the statement
+	// see https://medium.com/codechain/rust-underscore-does-not-bind-fec6a18115a8
+	let (_light_client_lock, _side_chain_lock) = EnclaveLock::write_all()?;
+
+	let mut validator = LightClientSeal::<PB>::unseal()?;
+
+	let authority = Ed25519Seal::unseal()?;
+	let state_key = AesSeal::unseal()?;
+
+	let rpc_author = GlobalAuthorContainer.get().ok_or_else(|| {
+		error!("Failed to retrieve author mutex. Maybe it's not initialized?");
+		Error::MutexAccess
+	})?;
+
+	let state_handler = Arc::new(GlobalFileStateHandler);
+	let stf_executor = Arc::new(StfExecutor::new(Arc::new(OcallApi), state_handler.clone()));
+
+	let latest_onchain_header = validator.latest_finalized_header(validator.num_relays()).unwrap();
+	let genesis_hash = validator.genesis_hash(validator.num_relays())?;
+	let extrinsics_factory =
+		ExtrinsicsFactory::new(genesis_hash, authority.clone(), GLOBAL_NONCE_CACHE.clone());
+
+	let top_pool_executor =
+		Arc::new(TopPoolOperationExecutor::<PB, SignedSidechainBlock, _, _>::new(
+			rpc_author.clone(),
+			stf_executor.clone(),
+		));
+
+	let block_composer =
+		Arc::new(BlockComposer::new(authority.clone(), state_key, rpc_author, stf_executor));
```

Comment:
All of this is also just constructing the necessary components. Will try to refactor this in the next PR (as described above)